bolts/fair/models/fair_mixup.py

```diff
@@ -0,0 +1,286 @@
+        lr_restart_mult: int = 2,
+        lr_sched_interval: SchedInterval = "epoch",
+        lr_sched_freq: int = 1,
+    ):
+        super().__init__()
+        self.enc = enc
+        self.clf = clf
+        self.net = nn.Sequential(self.enc, self.clf)
+        self._loss_fn = CrossEntropy()
+
+        self.mixup_lambda = mixup_lambda
+        self.alpha = alpha
+        self._target_name = "y"
+
+        self.learning_rate = lr
+        self.weight_decay = weight_decay
+        self.lr_initial_restart = lr_initial_restart
+        self.lr_restart_mult = lr_restart_mult
+        self.lr_sched_interval = lr_sched_interval
+        self.lr_sched_freq = lr_sched_freq
+
+        self.test_acc = torchmetrics.Accuracy()
+        self.train_acc = torchmetrics.Accuracy()
+        self.val_acc = torchmetrics.Accuracy()
+
+    @property
+    def target(self) -> str:
+        return self._target_name
+
+    @target.setter
+    def target(self, target: str) -> None:
+        self._target_name = target
+
+    def training_step(self, batch: DataBatch, batch_idx: int) -> STEP_OUTPUT:
+        inputs, x_a, x_b, targets_a, targets_b, _, _, lam, mixup_stats = self.mixup_data(
+            batch=batch, device=self.device, mix_lambda=self.mixup_lambda, alpha=self.alpha
+        )
+
+        logits = self.net(inputs)
+        loss_sup = self.mixup_criterion(
+            criterion=self._loss_fn, pred=logits, tgt_a=targets_a, tgt_b=targets_b, lam=lam
+        )
+
+        target = batch.y.view(-1).long()
+        self.train_acc(logits.argmax(-1), target)
+        self.log_dict(
+            {
+                f"train/loss": loss_sup.item(),
+                f"train/acc": self.train_acc,
+            }
+        )
+
+        ops = logits.sum()
+
+        # Smoothness Regularization
+        gradx = torch.autograd.grad(ops, inputs, create_graph=True)[0].view(inputs.shape[0], -1)
+        x_d = (x_b - x_a).view(inputs.shape[0], -1)
+        grad_inn = (gradx * x_d).sum(1).view(-1)
+        loss_grad = torch.abs(grad_inn.mean())
+
+        loss = loss_sup + lam * loss_grad
+
+        self.log_dict({"train/loss_sup": loss_sup.item(), "train/loss_mixup": loss_grad.item()})
+
+        return loss
+
+    def _get_loss(self, logits: Tensor, batch: DataBatch) -> Tensor:
+        return self._loss_fn(input=logits, target=batch.y)
+
+    def _inference_step(self, batch: DataBatch, stage: Stage) -> dict[str, Tensor]:
+        logits = self.net(batch.x)
+
+        loss = self._get_loss(logits, batch)
+        tm_acc = self.val_acc if stage == "validate" else self.test_acc
+        target = batch.y.view(-1).long()
+        _acc = tm_acc(logits.argmax(-1), target)
+        self.log_dict(
+            {
+                f"{stage}/loss": loss.item(),
+                f"{stage}/{self.target}_acc": _acc,
+            }
+        )
+
+        return {
+            "y": batch.y.view(-1),
+            "s": batch.s.view(-1),
+            "preds": logits.softmax(-1)[:, 1],
```

Comment:
Why do you index the first element of the first dimension here instead of just computing the argmax?

Reply:
because they want to report the mean prediction. I don't think we really care about this though